### PR TITLE
nmcli: allow IPv4/IPv6 configuration on ipip and sit devices

### LINKED ISSUE #3238: nmcli ignoring IP4 / IP6 settings on tunnel type sit
### Summary

IPV4/IPV6 settings are not getting configured on sit tunnel devices.
This bug exists due to IPV4 / IPV6 configuration is only applied if the device type is in an hardoded list of Ip supported devices and SIT is missing.

Relevant code:
https://github.com/ansible-collections/community.general/blob/6ac410b3f617074bb7af86050d8adfa4d495a3e6/plugins/modules/net_tools/nmcli.py#L821

Solution proposal:
Add connection type "sit" to the statically defined list of IP connection types.
I will open an Pull Request with the proposed change.

### Issue Type

Bug Report

### Component Name

nmcli

### Ansible Version

```console (paste below)
$ ansible --version
ansible [core 2.11.3]
  config file = None
  configured module search path = ['/home/opc/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.6/site-packages/ansible
  ansible collection location = /home/opc/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/local/bin/ansible
  python version = 3.6.8 (default, May 19 2021, 10:00:09) [GCC 8.4.1 20200928 (Red Hat 8.4.1-1.0.1)]
  jinja version = 2.10.1
  libyaml = True
```

### Configuration

_No response_

### OS / Environment

RHEL 8

### Steps to Reproduce

```
  nmcli:
    type: sit
    conn_name: sit_tb
    autoconnect: yes
    ip_tunnel_remote: 94.44.44.44
    ip6: 2:1:1:171::2/64
    gw6: 2:1:1:171::1
    method4: disabled
    state: present
```

### Expected Results

Tunnel should be configured with static IPv6 configurations and IPv4 disabled.

### Actual Results

Tunnel is configured with dhcp.

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by zerotens**
##### SUMMARY

Add connection type "sit" to the allowed connection types which allows configuring IPv4/IPv6 settings.

Fixes #3238 

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME

nmcli

##### ADDITIONAL INFORMATION

```paste below
  nmcli:
    type: sit
    conn_name: sit_tb
    autoconnect: yes
    ip_tunnel_remote: 94.44.44.44
    ip6: 2:1:1:171::2/64
    gw6: 2:1:1:171::1
    method4: disabled
    state: present
```
###### NetworkManager configuration file /etc/NetworkManager/system-connections/ before change

```
[connection]
id=sit_tb
uuid=dcae0d3a-9703-48c2-89c6-2a0f7881fe00
type=ip-tunnel
interface-name=sit_tb
permissions=

[ip-tunnel]
mode=3
remote=94.44.44.44

[ipv4]
dns-search=
method=auto

[ipv6]
addr-gen-mode=stable-privacy
dns-search=
method=auto

[proxy]
```

###### NetworkManager configuration file /etc/NetworkManager/system-connections/ after change

```
[connection]
id=sit_tb
uuid=7a21ffc6-0d31-40c1-bcef-825ce506a136
type=ip-tunnel
interface-name=sit_tb
permissions=
timestamp=1629390525

[ip-tunnel]
mode=3
remote=94.44.44.44

[ipv4]
dns-search=
method=disabled

[ipv6]
addr-gen-mode=stable-privacy
address1=2:1:1:171::2/64,2:1:1:171::1
dns-search=
method=manual

[proxy]
```

### CODE REVIEW

File: changelogs/fragments/3239-nmcli-sit-ip-config-bugfix.yaml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - "nmcli - Parse ip4/ip6 configuration arguments for sit tunnels (https://github.com/ansible-collections/community.general/pull/3239)."
```
Comment (Ajpantuso) [inline changelogs/fragments/3239-nmcli-sit-ip-config-bugfix.yaml:2]:
```suggestion
  - "nmcli - added ip4/ip6 configuration arguments for ``sit`` tunnels (https://github.com/ansible-collections/community.general/issues/3238, https://github.com/ansible-collections/community.general/pull/3239)."
```

**Review by Ajpantuso — CHANGES_REQUESTED** [review]
Not sure this is strictly a bugfix. It changes existing behavior and probably should be applied for `ipip` tunnels as well.

**Comment by zerotens** [comment]
> Not sure this is strictly a bugfix. It changes existing behavior and probably should be applied for `ipip` tunnels as well.

Hi, thanks for your feedback.

I declared it as an bug because nmcli.py does not throw an error if you declare IP4 / IP6 settings but those settings won't get applied.

And yes, the same logic should also be applied for `ipip` tunnels, but i am not familiar with `ipip` tunnels, so i decided to only change it for `sit` tunnels.

I also plan to add support for `gre` tunnels #3105.

**Comment by Ajpantuso** [comment]
Hmm, I think this is safe as a bugfix. The default values applied for ipv4/6 settings match `nmcli` and the unspecified values are discarded so the ultimate result should be the same. If you can demonstrate via regression testing that the existing behavior when no ipv4/6 related options are set is unchanged then this should be fine. I would include `ipip` tunnels as well though (and `gre` tunnels once that's implemented).

**Comment by zerotens** [comment]
> Hmm, I think this is safe as a bugfix. The default values applied for ipv4/6 settings match `nmcli` and the unspecified values are discarded so the ultimate result should be the same. If you can demonstrate via regression testing that the existing behavior when no ipv4/6 related options are set is unchanged then this should be fine. I would include `ipip` tunnels as well though (and `gre` tunnels once that's implemented).

Ok i will include `ipip` tunnels.

Although i am unsure how the regression test should work because a change for the generated command line will take place.

The default arguments won't be cleared and the resulting command is different, but the resulting configuration of nmcli will be the same. (Tested for nmcli tool, version 1.30.0-10.0.1.el8_4)

Old generated command: 
`/usr/bin/nmcli con add type ip-tunnel con-name test2 connection.interface-name test2 connection.autoconnect yes ip-tunnel.local 192.168.225.5 ip-tunnel.mode sit ip-tunnel.parent ens3 ip-tunnel.remote 192.168.225.6`

New generated command with my changes:
`/usr/bin/nmcli con add type ip-tunnel con-name test connection.interface-name test connection.autoconnect yes ipv4.dns '' ipv4.dns-search '' ipv4.ignore-auto-dns no ipv4.ignore-auto-routes no ipv4.routes '' ipv4.never-default no ipv4.may-fail yes ipv6.dns-search '' ipv6.ignore-auto-dns no ipv6.ignore-auto-routes no ip-tunnel.local 192.168.225.5 ip-tunnel.mode sit ip-tunnel.parent ens3 ip-tunnel.remote 192.168.225.6`

Old resulting configuration:
```
[connection]
id=test2
uuid=dc1ffa98-45ed-49f6-a7d1-167e762eef97
type=ip-tunnel
interface-name=test2
permissions=

[ip-tunnel]
local=192.168.225.5
mode=3
parent=ens3
remote=192.168.225.6

[ipv4]
dns-search=
method=auto

[ipv6]
addr-gen-mode=stable-privacy
dns-search=
method=auto

[proxy]
```

New resulting configuration:
```
[connection]
id=test
uuid=168ea1f9-7a96-41f3-a312-b5043888b797
type=ip-tunnel
interface-name=test
permissions=

[ip-tunnel]
local=192.168.225.5
mode=3
parent=ens3
remote=192.168.225.6

[ipv4]
dns-search=
method=auto

[ipv6]
addr-gen-mode=stable-privacy
dns-search=
method=auto

[proxy]
```

**Comment by Ajpantuso** [comment]
So regression should include any behavior from the perspective of an Ansible user (since the end result is unchanged). Basically if the task result is the same before and after this PR then it should be safe as a bugfix. You can either run the playbook with `-vvv` or print the result via `debug` and just check for any differences.

**Comment by zerotens** [comment]
> So regression should include any behavior from the perspective of an Ansible user (since the end result is unchanged). Basically if the task result is the same before and after this PR then it should be safe as a bugfix. You can either run the playbook with `-vvv` or print the result via `debug` and just check for any differences.

Hi, thanks for clarification.

I tested it a few times and everything behaves like it should be from the ansible user perspective.
Tests was done with the following playbook with the old and the new logic and relevant verbose module output is the same.

```
---
# file: test.yml
- name: Test
  hosts: servers
  tasks:
    - name: Test SIT
      become: true
      nmcli:
        type: sit
        conn_name: sit_test
        ifname: sit_test
        ip_tunnel_dev: ens3
        ip_tunnel_local: 192.168.225.5
        ip_tunnel_remote: 192.168.225.6
        state: present

    - name: Test IPIP
      become: true
      nmcli:
        type: ipip
        conn_name: ipip_test
        ifname: ipip_test
        ip_tunnel_dev: ens3
        ip_tunnel_local: 192.168.225.5
        ip_tunnel_remote: 192.168.225.6
        state: present
```

**Review by Ajpantuso — APPROVED** [review]
LGTM, thanks! @hummeltech would you like to give this a look as well given your recent work on this module?

**Review by hummeltech — APPROVED** [review]
Thanks, yes this looks good to me too, I also tested it out without issue.

**Comment by felixfontein** [comment]
@zerotens thanks for implementing this!
@Ajpantuso @hummeltech thanks for reviewing!

**Pull request merged**